services/report/languages/pycoverage.py

```diff
@@ -47,12 +78,20 @@ def process(
                     (COVERAGE_HIT, ln) for ln in file_coverage["executed_lines"]
                 ] + [(COVERAGE_MISS, ln) for ln in file_coverage["missing_lines"]]
                 for cov, ln in lines_and_coverage:
-                    label_list_of_lists = [
-                        [self._convert_testname_to_label(testname, labels_table)]
-                        for testname in file_coverage.get("contexts", {}).get(
-                            str(ln), []
-                        )
-                    ]
+                    if report_builder_session.should_use_label_index:
+                        label_list_of_lists = [
+                            self._get_list_of_label_ids(
+                                report_builder_session.label_index,
+                                file_coverage.get("contexts", {}).get(str(ln), []),
+                            )
+                        ]
+                    else:
+                        label_list_of_lists = [
+                            [self._convert_testname_to_label(testname)]
+                            for testname in file_coverage.get("contexts", {}).get(
+                                str(ln), []
+                            )
+                        ]
```

Comment:
top branch creates an outer list containing a single large inner list, bottom branch creates an outer list containing many single-entry inner lists. is it okay that they don't match?

Reply:
THat's a very good catch. There is a difference indeed.

This goes to `ReportBuilderSession.create_coverage_line` ([here](https://github.com/codecov/worker/blob/gio/label-compressed/update-shared/services/report/report_builder.py#L202)). On the top it creates a single datapoint with all labels. In the bottom it creates many datapoints, one label per datapoint.

This difference is crucial when updating the report with new uploads. In particular when we [delete datapoints](https://github.com/codecov/shared/blob/gio/label-lookup/shared/reports/editable.py#L52). It seems that the bottom one is more precise and keeps data available longer. Sadly it also uses more space. But I'll keep the same way it was before.